### PR TITLE
[JENKINS-44860] `UsernameCredentials.isUsernameSecret`

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-44860](https://issues.jenkins.io/browse/JENKINS-44860)

https://github.com/jenkinsci/credentials-binding-plugin/pull/131#issuecomment-832854838

### CODE REVIEW

File: src/main/java/com/cloudbees/plugins/credentials/CredentialsNameProvider.java [hunk 1]
```diff
@@ -90,7 +90,11 @@ Result name(@NonNull Credentials credentials, @NonNull Class<?> clazz) {
         if (nameWith != null) {
             try {
                 CredentialsNameProvider nameProvider = nameWith.value().newInstance();
-                return new Result(nameProvider.getName(credentials), nameWith.priority());
+                String name = nameProvider.getName(credentials);
+                if (!name.isEmpty()) {
```
Comment (jglick) [inline src/main/java/com/cloudbees/plugins/credentials/CredentialsNameProvider.java:94]: Make it possible to just `return credentials.getDescription()` and if that happens to be the empty string, still behave reasonably (choose a different provider).

File: src/main/java/com/cloudbees/plugins/credentials/impl/UsernamePasswordCredentialsImpl.java [hunk 2]
```diff
@@ -53,6 +55,9 @@
     @NonNull
     private final Secret password;
 
+    @Nullable
+    private Boolean usernameSecret = false;
```
Comment (jglick) [inline src/main/java/com/cloudbees/plugins/credentials/impl/UsernamePasswordCredentialsImpl.java:59]: Default false in the Java constructor which would be used only from tests or scripts.

File: src/main/java/com/cloudbees/plugins/credentials/impl/UsernamePasswordCredentialsImpl.java [hunk 3]
```diff
@@ -72,6 +77,13 @@ public UsernamePasswordCredentialsImpl(@CheckForNull CredentialsScope scope,
         this.password = Secret.fromString(password);
     }
 
+    private Object readResolve() {
+        if (usernameSecret == null) {
+            usernameSecret = true;
```
Comment (jglick) [inline src/main/java/com/cloudbees/plugins/credentials/impl/UsernamePasswordCredentialsImpl.java:82]: Default true for deserialized credentials, either from settings predating the upgrade, or REST uploads as XML.

File: src/main/resources/com/cloudbees/plugins/credentials/impl/UsernamePasswordCredentialsImpl/credentials.jelly [hunk 4]
```diff
@@ -28,6 +28,9 @@
   <f:entry title="${%Username}" field="username">
     <f:textbox/>
   </f:entry>
+  <f:entry field="usernameSecret">
+    <f:checkbox title="${%Treat username as secret}"/>
```
Comment (jglick) [inline src/main/resources/com/cloudbees/plugins/credentials/impl/UsernamePasswordCredentialsImpl/credentials.jelly:32]: Default false for new credentials created from the GUI.

**Review by Wadeck — APPROVED** [review]
Manually tested:
- ✔️ Existing credentials are not changing behavior
- ✔️ New credentials let the checkbox uncheck (but this could be changed)

Of course, no masking until credentials-masking is merged as well.

Will merge this for next release

**Comment by bitwiseman** [comment]
Why not default to `usernameSecret=true`  in all cases?

**Comment by jglick** [comment]
@bitwiseman https://github.com/jenkinsci/credentials-plugin/pull/205#discussion_r627811232 for the GUI.

**Review by car-roll — APPROVED** [review]
I'm fine with making username not secret as default going forward since the behavior for legacy credentials is to keep them hidden.

**Comment by bitwiseman** [comment]
> @bitwiseman https://github.com/jenkinsci/credentials-plugin/pull/205#discussion_r627811232 for the GUI.

I don't understand. Yes, I know it defaults to false in GUI. Why does it default to false? The GUI couldn't default to true/checked?

**Comment by jglick** [comment]
> I know it defaults to false in GUI. Why does it default to false?

Sorry, did not get your question. We expect the username to be nonsecret in most cases, and until now it was displayed in credentials dropdowns to anyone with view configuration permissions. In fact the originally submitted PRs just _unconditionally_ began showing usernames in logs, merely by virtue of updating the `credentials-binding` plugin, which I was uncomfortable with. So here I am trying to walk a compromise path that poses no risk of unwanted surprise for an existing installation, while guiding new configurations to display the username.

File: pom.xml [hunk 5]
```diff
@@ -65,7 +65,7 @@
   </scm>
 
   <properties>
-    <revision>2.4.2</revision>
+    <revision>2.5</revision>
```
Comment (jglick) [inline pom.xml:68]: Just proposing a version bump. Do we really need to keep the micro version? Can revert if desired.

**Comment by bitwiseman** [comment]
> > I know it defaults to false in GUI. Why does it default to false?
> 
> Sorry, did not get your question. We expect the username to be nonsecret in most cases, and until now it was displayed in credentials dropdowns to anyone with view configuration permissions. In fact the originally submitted PRs just _unconditionally_ began showing usernames in logs, merely by virtue of updating the `credentials-binding` plugin, which I was uncomfortable with. So here I am trying to walk a compromise path that poses no risk of unwanted surprise for an existing installation, while guiding new configurations to display the username.

Understood. Sounds good, thanks for the explanation.

**Review by bitwiseman — APPROVED** [review]

**Comment by jglick** [comment]
@bitwiseman @car-roll @Wadeck any updates on this & https://github.com/jenkinsci/credentials-binding-plugin/pull/132 & https://github.com/jenkinsci/ssh-credentials-plugin/pull/70?

**Review by bitwiseman — APPROVED** [review]

**Pull request merged**